Bio/Affy/CelFile.py

```diff
@@ -78,8 +94,204 @@ def __init__(self):
+
+    Most importantly record.intensities correspond to intensities from the CEL file.
+
+    record.mask and record.outliers are not set.
+
+    Example Usage:
+
+    >>> from Bio.Affy import CelFile
+    >>> with open("Affy/affy_v4_example.CEL", "rb") as handle:
+    ...     c = CelFile.read_v4(handle)
+    ...
+    >>> c.version == 4
+    True
+    >>> print(c.intensities.shape)
+    (5, 5)
     """
-    Read the information in a cel file, and store it in a Record.
+    # We follow the documentation here:
+    # http://www.affymetrix.com/estore/support/developer/powertools/changelog/gcos-agcc/cel.html.affx
+    record = Record()
+    preHeaders = ["magic", "version", "columns", "rows", "cellNo", "headerLen"]
+    preHeadersMap = dict()
+    headersMap = dict()
+
+    # load pre-headers
+    for name in preHeaders:
+        preHeadersMap[name] = struct.unpack("<i", f.read(4))[0]
+    char = f.read(preHeadersMap["headerLen"])
+    header = char.decode("ascii", "ignore")
+    for header in header.split("\n"):
+        if "=" in header:
+            header = header.split("=")
+            headersMap[header[0]] = "=".join(header[1:])
+
+    # for debugging
+    # pp.pprint("preHeadersMap")
+    # pp.pprint(preHeadersMap)
+    # pp.pprint("headersMap")
+    # pp.pprint(headersMap)
+
+    record.version = preHeadersMap["version"]
+    if record.version != 4:
+        raise ValueError("You are trying to parse CEL file version 4. This file violates the structure expected from CEL file version 4")
+    record.GridCornerUL = headersMap["GridCornerUL"]
+    record.GridCornerUR = headersMap["GridCornerUR"]
+    record.GridCornerLR = headersMap["GridCornerLR"]
+    record.GridCornerLL = headersMap["GridCornerLL"]
+    record.DatHeader = headersMap["DatHeader"]
+    record.Algorithm = headersMap["Algorithm"]
+    record.AlgorithmParameters = headersMap["AlgorithmParameters"]
+    record.NumberCells = preHeadersMap["cellNo"]
+    # record.intensities are set below
+    # record.stdevs are set below
+    # record.npix are set below
+    record.nrows = int(headersMap["Rows"])
+    record.ncols = int(headersMap["Cols"])
+
+    # These cannot be reliably set in v4, because of discrepancies between real
+    # data and the documented format.
+    record.nmask = None
+    record.mask = None
+    record.noutliers = None
+    record.outliers = None
+    record.modified = None
+
+    # Real data never seems to have anything but zeros here, but we don't want
+    # to take chances. Raising an error is better than returning unreliable
+    # data.
+    def raiseBadHeader(field, expected):
+        actual = int(headersMap[field])
+        message = "The header {field} is expected to be 0, not {value}".format(value=actual, field=field)
+        if actual != expected:
+            raise ValueError(message)
+
+    raiseBadHeader("Axis-invertX", 0)
+
+    raiseBadHeader("AxisInvertY", 0)
```

Comment:
Is it really ``Axis-invertX`` (with a hyphen/minus) and ``AxisInvertY`` (without)?